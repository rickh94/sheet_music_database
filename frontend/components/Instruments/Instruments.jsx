import React, { Component } from 'react'
import PropTypes from 'prop-types'

import Header from '../Header'
import Container from 'react-bulma-components/lib/components/container'
import Box from 'react-bulma-components/lib/components/box'
import Level from 'react-bulma-components/lib/components/level'
import Heading from 'react-bulma-components/lib/components/heading'
import ListItem from '../ListItem'

export class Instruments extends Component {
  state = {
    errors: {}
  }

  static propTypes = {
    token: PropTypes.string,
    history: PropTypes.object.isRequired,
    alert: PropTypes.object.isRequired,
    instruments: PropTypes.object.isRequired
  }

  render() {
    const instrumentList = this.props.instruments.list
    return (
      <React.Fragment>
        <Header />
        <Container>
          <Box className="margin-default">
            <Level>
              <Heading size={3} className="absolutely-no-margin level-item level-left">
                Instruments
              </Heading>
            </Level>
            {instrumentList.map(instrument => (
              <ListItem
                item={instrument}
                key={instrument.id}
                saveCallback={() => {}}
                errors={this.state.errors[instrument.id]}
              />
            ))}
          </Box>
        </Container>
      </React.Fragment>
    )
  }
}

export default Instruments
