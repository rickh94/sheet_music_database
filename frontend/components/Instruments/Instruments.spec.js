import React from 'react'
import { shallow } from 'enzyme'
import { Instruments } from './Instruments'
import '../../setupTests'
import {
  testToken,
  testLoggedIn,
  testGetData,
  testGetDataFailed,
  clickButton
} from '../../testHelpers'

describe('<Instruments />', () => {
  let wrapper
  beforeEach(() => {
    wrapper = shallow(
      <Instruments
        history={{ push: jest.fn() }}
        token={testToken}
        alert={{ show: jest.fn() }}
        getInstruments={jest.fn()}
        createInstrument={jest.fn()}
        deleteInstrument={jest.fn()}
        updateInstrument={jest.fn()}
        instruments={{ list: [] }}
      />
    )
  })

  it('matches the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('renders the header', () => {
    expect(wrapper.exists('Connect(Header)')).toBeTruthy()
  })

  it('renders a container', () => {
    expect(wrapper.exists('Container')).toBeTruthy()
  })

  it('renders the instruments from the list', () => {
    const instruments = { list: [{ id: 1, name: 'violin' }, { id: 2, name: 'viola' }] }
    wrapper.setProps({ instruments })
    expect(wrapper.find('ListItem').length).toEqual(2)
  })

  it('redirects if not logged in', () => {
    testLoggedIn(wrapper)
  })

  it('gets instruments if logged in', () => {
    testGetData(wrapper, 'getInstruments')
  })

  it('shows an error on get failure', () => {
    testGetDataFailed(wrapper, 'getInstruments', 'Instruments')
  })

  it('renders a create button', () => {
    wrapper.find('#activate-create').simulate('click')
    expect(wrapper.state().createMode).toBeTruthy()
  })

  it('renders a modal to create new instruments', () => {
    expect(wrapper.exists('Modal')).toBeTruthy()
    expect(wrapper.find('Modal').contains('New Instrument')).toBeTruthy()
    expect(wrapper.find('Modal').exists('TextFieldWithErrors')).toBeTruthy()
  })

  it('has a button to save created instrument, and resets state', () => {
    const createInstrument = jest.fn()
    const newInstrumentName = 'new test instrument name'
    const spy = jest.spyOn(wrapper.instance(), 'handleSubmit')
    wrapper.setProps({ createInstrument })
    wrapper.setState({ newInstrumentName, createMode: true })
    clickButton(wrapper, 'Save')
    expect(spy).toHaveBeenCalled()
  })

  it('has a cancel button that closes the modal and reset newinstrumentname', () => {
    wrapper.setState({ createMode: true, newInstrumentName: 'test instrument name' })
    clickButton(wrapper, 'Cancel')
    expect(wrapper.state().createMode).toBeFalsy()
    expect(wrapper.state().newInstrumentName).toEqual('')
  })

  describe('handle submit', () => {
    const e = { preventDefault: jest.fn() }
    it('returns and sets errors if no name', async () => {
      const createInstrument = jest.fn()
      wrapper.setProps({ createInstrument })
      wrapper.setState({ newInstrumentName: '', createMode: true })
      await wrapper.instance().handleSubmit(e)
      expect(createInstrument).not.toHaveBeenCalled()
      expect(wrapper.state().errors.name).toEqual('Instrument name is required')
      expect(wrapper.state().createMode).toBeTruthy()
    })

    it('calls create instrument and clears form on success', async () => {
      const createInstrument = jest.fn()
      const show = jest.fn()
      createInstrument.mockReturnValue(true)
      wrapper.setProps({ createInstrument, alert: { show } })
      wrapper.setState({
        newInstrumentName: 'testname', createMode: true
      })
      await wrapper.instance().handleSubmit(e)
      expect(createInstrument).toHaveBeenCalledWith(testToken, 'testname')
      expect(show).toHaveBeenCalledWith(<span className="alert-text">Instrument Created</span>)
      expect(wrapper.state().newInstrumentName).toEqual('')
    })

    it('sets errors and shows alert on failure', async () => {
      const createInstrument = jest.fn()
      const show = jest.fn()
      createInstrument.mockReturnValue(false)
      wrapper.setProps({
        createInstrument,
        alert: { show },
        instruments: { errors: { name: 'test error' }, list: [] }
      })
      wrapper.setState({
        newInstrumentName: 'testname', createMode: true
      })
      await wrapper.instance().handleSubmit(e)
      expect(createInstrument).toHaveBeenCalledWith(testToken, 'testname')
      expect(show).toHaveBeenCalledWith(
        <span className="alert-text">Instrument Creation Failed</span>,
        { type: 'error' }
      )
      expect(wrapper.state().errors.name).toEqual('test error')
      expect(wrapper.state().createMode).toBeTruthy()
    })
  })
})
