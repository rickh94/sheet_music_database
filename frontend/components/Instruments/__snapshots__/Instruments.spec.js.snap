// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Instruments /> matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Instruments
    alert={
      Object {
        "list": Array [],
      }
    }
    createInstrument={[MockFunction]}
    deleteInstrument={[MockFunction]}
    getInstruments={[MockFunction]}
    history={
      Object {
        "push": [MockFunction],
      }
    }
    instruments={
      Object {
        "list": Array [],
      }
    }
    token="testtoken"
    updateInstrument={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Connect(Header) />,
        <Container
          breakpoint={null}
          fluid={false}
          renderAs="div"
          style={Object {}}
        >
          <Box
            className="margin-default"
            renderAs="div"
            style={Object {}}
          >
            <Level
              breakpoint={null}
              className=""
              renderAs="div"
              style={Object {}}
            >
              <Heading
                className="absolutely-no-margin level-item level-left"
                heading={false}
                renderAs="h1"
                size={3}
                spaced={false}
                subtitle={false}
                weight={null}
              >
                Instruments
              </Heading>
            </Level>
          </Box>
        </Container>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "backgroundColor": undefined,
          "breakpoint": null,
          "children": <Box
            className="margin-default"
            renderAs="div"
            style={Object {}}
          >
            <Level
              breakpoint={null}
              className=""
              renderAs="div"
              style={Object {}}
            >
              <Heading
                className="absolutely-no-margin level-item level-left"
                heading={false}
                renderAs="h1"
                size={3}
                spaced={false}
                subtitle={false}
                weight={null}
              >
                Instruments
              </Heading>
            </Level>
          </Box>,
          "className": undefined,
          "clearfix": undefined,
          "clipped": undefined,
          "fluid": false,
          "hidden": undefined,
          "invisible": undefined,
          "italic": undefined,
          "marginless": undefined,
          "overlay": undefined,
          "paddingless": undefined,
          "pull": undefined,
          "radiusless": undefined,
          "renderAs": "div",
          "responsive": undefined,
          "shadowless": undefined,
          "style": Object {},
          "textAlignment": undefined,
          "textColor": undefined,
          "textSize": undefined,
          "textTransform": undefined,
          "textWeight": undefined,
          "unselectable": undefined,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "backgroundColor": undefined,
            "children": Array [
              <Level
                breakpoint={null}
                className=""
                renderAs="div"
                style={Object {}}
              >
                <Heading
                  className="absolutely-no-margin level-item level-left"
                  heading={false}
                  renderAs="h1"
                  size={3}
                  spaced={false}
                  subtitle={false}
                  weight={null}
                >
                  Instruments
                </Heading>
              </Level>,
              Array [],
            ],
            "className": "margin-default",
            "clearfix": undefined,
            "clipped": undefined,
            "hidden": undefined,
            "invisible": undefined,
            "italic": undefined,
            "marginless": undefined,
            "overlay": undefined,
            "paddingless": undefined,
            "pull": undefined,
            "radiusless": undefined,
            "renderAs": "div",
            "responsive": undefined,
            "shadowless": undefined,
            "style": Object {},
            "textAlignment": undefined,
            "textColor": undefined,
            "textSize": undefined,
            "textTransform": undefined,
            "textWeight": undefined,
            "unselectable": undefined,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "backgroundColor": undefined,
                "breakpoint": null,
                "children": <Heading
                  className="absolutely-no-margin level-item level-left"
                  heading={false}
                  renderAs="h1"
                  size={3}
                  spaced={false}
                  subtitle={false}
                  weight={null}
                >
                  Instruments
                </Heading>,
                "className": "",
                "clearfix": undefined,
                "clipped": undefined,
                "hidden": undefined,
                "invisible": undefined,
                "italic": undefined,
                "marginless": undefined,
                "overlay": undefined,
                "paddingless": undefined,
                "pull": undefined,
                "radiusless": undefined,
                "renderAs": "div",
                "responsive": undefined,
                "shadowless": undefined,
                "style": Object {},
                "textAlignment": undefined,
                "textColor": undefined,
                "textSize": undefined,
                "textTransform": undefined,
                "textWeight": undefined,
                "unselectable": undefined,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "backgroundColor": undefined,
                  "children": "Instruments",
                  "className": "absolutely-no-margin level-item level-left",
                  "clearfix": undefined,
                  "clipped": undefined,
                  "heading": false,
                  "hidden": undefined,
                  "invisible": undefined,
                  "italic": undefined,
                  "marginless": undefined,
                  "overlay": undefined,
                  "paddingless": undefined,
                  "pull": undefined,
                  "radiusless": undefined,
                  "renderAs": "h1",
                  "responsive": undefined,
                  "shadowless": undefined,
                  "size": 3,
                  "spaced": false,
                  "subtitle": false,
                  "textAlignment": undefined,
                  "textColor": undefined,
                  "textSize": undefined,
                  "textTransform": undefined,
                  "textWeight": undefined,
                  "unselectable": undefined,
                  "weight": null,
                },
                "ref": null,
                "rendered": "Instruments",
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Connect(Header) />,
          <Container
            breakpoint={null}
            fluid={false}
            renderAs="div"
            style={Object {}}
          >
            <Box
              className="margin-default"
              renderAs="div"
              style={Object {}}
            >
              <Level
                breakpoint={null}
                className=""
                renderAs="div"
                style={Object {}}
              >
                <Heading
                  className="absolutely-no-margin level-item level-left"
                  heading={false}
                  renderAs="h1"
                  size={3}
                  spaced={false}
                  subtitle={false}
                  weight={null}
                >
                  Instruments
                </Heading>
              </Level>
            </Box>
          </Container>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "backgroundColor": undefined,
            "breakpoint": null,
            "children": <Box
              className="margin-default"
              renderAs="div"
              style={Object {}}
            >
              <Level
                breakpoint={null}
                className=""
                renderAs="div"
                style={Object {}}
              >
                <Heading
                  className="absolutely-no-margin level-item level-left"
                  heading={false}
                  renderAs="h1"
                  size={3}
                  spaced={false}
                  subtitle={false}
                  weight={null}
                >
                  Instruments
                </Heading>
              </Level>
            </Box>,
            "className": undefined,
            "clearfix": undefined,
            "clipped": undefined,
            "fluid": false,
            "hidden": undefined,
            "invisible": undefined,
            "italic": undefined,
            "marginless": undefined,
            "overlay": undefined,
            "paddingless": undefined,
            "pull": undefined,
            "radiusless": undefined,
            "renderAs": "div",
            "responsive": undefined,
            "shadowless": undefined,
            "style": Object {},
            "textAlignment": undefined,
            "textColor": undefined,
            "textSize": undefined,
            "textTransform": undefined,
            "textWeight": undefined,
            "unselectable": undefined,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "backgroundColor": undefined,
              "children": Array [
                <Level
                  breakpoint={null}
                  className=""
                  renderAs="div"
                  style={Object {}}
                >
                  <Heading
                    className="absolutely-no-margin level-item level-left"
                    heading={false}
                    renderAs="h1"
                    size={3}
                    spaced={false}
                    subtitle={false}
                    weight={null}
                  >
                    Instruments
                  </Heading>
                </Level>,
                Array [],
              ],
              "className": "margin-default",
              "clearfix": undefined,
              "clipped": undefined,
              "hidden": undefined,
              "invisible": undefined,
              "italic": undefined,
              "marginless": undefined,
              "overlay": undefined,
              "paddingless": undefined,
              "pull": undefined,
              "radiusless": undefined,
              "renderAs": "div",
              "responsive": undefined,
              "shadowless": undefined,
              "style": Object {},
              "textAlignment": undefined,
              "textColor": undefined,
              "textSize": undefined,
              "textTransform": undefined,
              "textWeight": undefined,
              "unselectable": undefined,
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "backgroundColor": undefined,
                  "breakpoint": null,
                  "children": <Heading
                    className="absolutely-no-margin level-item level-left"
                    heading={false}
                    renderAs="h1"
                    size={3}
                    spaced={false}
                    subtitle={false}
                    weight={null}
                  >
                    Instruments
                  </Heading>,
                  "className": "",
                  "clearfix": undefined,
                  "clipped": undefined,
                  "hidden": undefined,
                  "invisible": undefined,
                  "italic": undefined,
                  "marginless": undefined,
                  "overlay": undefined,
                  "paddingless": undefined,
                  "pull": undefined,
                  "radiusless": undefined,
                  "renderAs": "div",
                  "responsive": undefined,
                  "shadowless": undefined,
                  "style": Object {},
                  "textAlignment": undefined,
                  "textColor": undefined,
                  "textSize": undefined,
                  "textTransform": undefined,
                  "textWeight": undefined,
                  "unselectable": undefined,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "backgroundColor": undefined,
                    "children": "Instruments",
                    "className": "absolutely-no-margin level-item level-left",
                    "clearfix": undefined,
                    "clipped": undefined,
                    "heading": false,
                    "hidden": undefined,
                    "invisible": undefined,
                    "italic": undefined,
                    "marginless": undefined,
                    "overlay": undefined,
                    "paddingless": undefined,
                    "pull": undefined,
                    "radiusless": undefined,
                    "renderAs": "h1",
                    "responsive": undefined,
                    "shadowless": undefined,
                    "size": 3,
                    "spaced": false,
                    "subtitle": false,
                    "textAlignment": undefined,
                    "textColor": undefined,
                    "textSize": undefined,
                    "textTransform": undefined,
                    "textWeight": undefined,
                    "unselectable": undefined,
                    "weight": null,
                  },
                  "ref": null,
                  "rendered": "Instruments",
                  "type": [Function],
                },
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
